flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/KinesisStreamsSource.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.flink.connector.kinesis.source.enumerator.KinesisStreamsSourceEnumeratorStateSerializer;
+import org.apache.flink.connector.kinesis.source.proxy.KinesisStreamProxy;
+import org.apache.flink.connector.kinesis.source.reader.KinesisStreamsRecordEmitter;
+import org.apache.flink.connector.kinesis.source.reader.KinesisStreamsSourceReader;
+import org.apache.flink.connector.kinesis.source.reader.PollingKinesisShardSplitReader;
+import org.apache.flink.connector.kinesis.source.serialization.KinesisDeserializationSchema;
+import org.apache.flink.connector.kinesis.source.split.KinesisShardSplit;
+import org.apache.flink.connector.kinesis.source.split.KinesisShardSplitSerializer;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.UserCodeClassLoader;
+
+import software.amazon.awssdk.http.SdkHttpClient;
+import software.amazon.awssdk.http.apache.ApacheHttpClient;
+import software.amazon.awssdk.services.kinesis.KinesisClient;
+import software.amazon.awssdk.services.kinesis.model.Record;
+import software.amazon.awssdk.utils.AttributeMap;
+
+import java.util.Properties;
+import java.util.function.Supplier;
+
+/**
+ * The {@link KinesisStreamsSource} is an exactly-once parallel streaming data source that
+ * subscribes to a single AWS Kinesis data stream. It is able to handle resharding of streams, and
+ * stores its current progress in Flink checkpoints. The source will read in data from the Kinesis
+ * Data stream, deserialize it using the provided {@link DeserializationSchema}, and emit the record
+ * into the Flink job graph.
+ *
+ * <p>Exactly-once semantics. To leverage Flink's checkpointing mechanics for exactly-once stream
+ * processing, the Kinesis Source is implemented with the AWS Java SDK, instead of the officially
+ * recommended AWS Kinesis Client Library. The source will store its current progress in Flink
+ * checkpoint/savepoint, and will pick up from where it left off upon restore from the
+ * checkpoint/savepoint.
+ *
+ * <p>Initial starting points. The Kinesis Streams Source supports reads starting from TRIM_HORIZON,
+ * LATEST, and AT_TIMESTAMP.
+ *
+ * @param <T> the data type emitted by the source
+ */
+@Experimental
+public class KinesisStreamsSource<T>
+        implements Source<T, KinesisShardSplit, KinesisStreamsSourceEnumeratorState> {
+
+    private final String streamArn;
+    private final Properties consumerConfig;
+    private final KinesisDeserializationSchema<T> deserializationSchema;
+    private final KinesisShardAssigner kinesisShardAssigner;
+
+    KinesisStreamsSource(
+            String streamArn,
+            Properties consumerConfig,
+            KinesisDeserializationSchema<T> deserializationSchema,
+            KinesisShardAssigner kinesisShardAssigner) {
+        Preconditions.checkNotNull(streamArn);
```

Comment:
Do we want to do a notEmpty check on the stream ARN?

Reply:
sure